src/endpoint.rs

```diff
@@ -14,23 +14,141 @@
  *  limitations under the License.
  */
 
-// TODO Move endpoint.rs out of config/ into cluster/
-use crate::cluster::Endpoint;
-use std::sync::Arc;
+//! Types representing where the data is the sent.
 
-#[derive(Debug)]
-pub struct EmptyListError;
+use std::{net::SocketAddr, sync::Arc};
 
-#[derive(Debug)]
-pub struct AllEndpointsRemovedError;
+use serde::{Deserialize, Serialize};
 
-#[derive(Debug)]
-pub struct IndexOutOfRangeError;
+use crate::metadata::Metadata;
 
-/// Endpoints represents the set of all known upstream endpoints.
+/// Represents the set of all known upstream endpoints.
 #[derive(Clone, Debug, PartialEq)]
 pub struct Endpoints(Arc<Vec<Endpoint>>);
 
+/// A destination endpoint with any associated metadata.
+#[derive(Debug, Deserialize, Serialize, PartialEq, Clone, PartialOrd, Eq)]
+#[non_exhaustive]
+#[serde(deny_unknown_fields)]
+pub struct Endpoint {
+    pub address: SocketAddr,
+    #[serde(default)]
+    pub metadata: Metadata<EndpointMetadata>,
+}
+
+impl Endpoint {
+    /// Creates a new [`Endpoint`] with no metadata.
+    pub fn new(address: SocketAddr) -> Self {
+        Self::with_metadata(address, Metadata::default())
+    }
+
+    /// Creates a new [`Endpoint`] with the specified `metadata`.
+    pub fn with_metadata(
+        address: SocketAddr,
+        metadata: impl Into<Metadata<EndpointMetadata>>,
+    ) -> Self {
+        Self {
+            address,
+            metadata: metadata.into(),
+        }
+    }
+}
+
+impl Default for Endpoint {
+    fn default() -> Self {
+        Self {
+            address: std::net::SocketAddrV6::new(std::net::Ipv6Addr::UNSPECIFIED, 0, 0, 0).into(),
+            metadata: <_>::default(),
+        }
+    }
+}
+
+/// Metadata specific to endpoints.
+#[derive(Default, Debug, Deserialize, Serialize, PartialEq, Clone, PartialOrd, Eq)]
+#[non_exhaustive]
+pub struct EndpointMetadata {
+    #[serde(with = "base64_set")]
+    pub tokens: base64_set::Set,
```

Comment:
The rest of the dynamic metadata seems to be missing?

Reply:
Dynamic metadata is stored in `Metadata<T>`.